smarts/sstudio/genhistories.py

```diff
@@ -450,6 +460,159 @@ def column_val_in_row(self, row, col_name):
+                record_data = f.read(record_len)
+                _ = f.read(4)  # masked_crc32_of_data (ignore)
+                yield record_data
+
+    @property
+    def rows(self):
+        def lerp(a, b, t):
+            return t * (b - a) + a
+
+        if "scenario_id" not in self._dataset_spec:
+            errmsg = "Dataset spec requires scenario_id to be set"
+            self._log.error(errmsg)
+            raise ValueError(errmsg)
+        scenario_id = self._dataset_spec["scenario_id"]
+
+        # Loop over the scenarios in the TFRecord and check its ID for a match
+        scenario = None
+        dataset = Waymo.read_dataset(self._dataset_spec["input_path"])
+        for record in dataset:
+            parsed_scenario = scenario_pb2.Scenario()
+            parsed_scenario.ParseFromString(bytearray(record))
+            if parsed_scenario.scenario_id == scenario_id:
+                scenario = parsed_scenario
+                break
+
+        if scenario == None:
+            errmsg = f"Dataset file does not contain scenario with id: {scenario_id}"
+            self._log.error(errmsg)
+            raise ValueError(errmsg)
+
+        for i in range(len(scenario.tracks)):
+            vehicle_id = scenario.tracks[i].id
+            vehicle_type = self._lookup_agent_type(scenario.tracks[i].object_type)
+            rows = []
+
+            # First pass -- extract data
+            for j in range(len(scenario.timestamps_seconds)):
+                obj_state = scenario.tracks[i].states[j]
+                vel = np.array([obj_state.velocity_x, obj_state.velocity_y])
+
+                row = {}
+                row["valid"] = obj_state.valid
+                row["vehicle_id"] = vehicle_id
+                row["type"] = vehicle_type
+                row["length"] = obj_state.length
+                row["height"] = obj_state.height
+                row["width"] = obj_state.width
+                row["sim_time"] = scenario.timestamps_seconds[j]
+                row["position_x"] = obj_state.center_x
+                row["position_y"] = obj_state.center_y
+                row["heading_rad"] = obj_state.heading - math.pi / 2
+                row["speed"] = np.linalg.norm(vel)
+                row["lane_id"] = 0
+                row["is_ego_vehicle"] = 1 if i == scenario.sdc_track_index else 0
+                rows.append(row)
+
+            # Second pass -- align timesteps to 10 Hz and interpolate trajectory data if needed
+            for j in range(len(scenario.timestamps_seconds)):
+                row = rows[j]
+                if row["valid"] == False:
+                    continue
+                timestep = 0.1
+                time_current = row["sim_time"]
+                time_expected = round(j * timestep, 3)
+                time_error = time_current - time_expected
+
+                if time_error == 0:
+                    continue
+
+                # Align to expected timestep
+                row["sim_time"] = time_expected
```

Comment:
In general, for all of the `row` updates that follow, since you're replacing `row[j][...]` immediately, if you have to interpolate the _next_ value based on this one, on so on, errors could compound.  It might be better to only update the `row[j][...]` entries after we've interpolated everything.  (I realize it's not as convenient in terms of code though.)